Bug: Type checker silently accepts single-variable assignment from multi-return functions, causing misleading errors

## Description

When assigning a multi-return function result to a single variable, the type checker silently registers the variable with an empty type instead of emitting an error. This causes confusing downstream errors when the variable is used.

## Example

```ez
import @db, @maps

do example() {
    temp database, _ = db.open("test.ezdb")
    temp value = db.get(database, "key")  // db.get returns (string, bool)
    
    temp m map[string:string] = {}
    maps.set(m, "key", value)  // Error here, not at the assignment
}
```

**Current error (misleading):**
```
error[E3001]: maps.set: value type mismatch - map has value type string, but got
  --> file.ez:7:25
   |
 7 |         maps.set(m, "key", value)
   |                         ^ types do not match
```

Note: "but got" shows nothing because `value` has an empty type.

**Expected error:**
```
error: cannot assign multi-return function result to single variable
  --> file.ez:4:17
   |
 4 |     temp value = db.get(database, "key")
   |                  ^ db.get() returns (string, bool), use: temp value, _ = db.get(...)
```

## Root Cause

In `pkg/typechecker/typechecker.go`:

1. **Line 5130-5134** in `inferModuleCallType`:
```go
if returnTypes := tc.getModuleMultiReturnTypes(resolvedModuleName, funcName, args); len(returnTypes) > 1 {
    // Multi-return stdlib calls cannot be inferred as a single value.
    return "", true  // Returns empty string but ok=true
}
```

2. **Lines 1808-1826** in `checkVariableDeclaration`:
```go
if declaredType == "" {
    inferredType, ok := tc.inferExpressionType(decl.Value)
    if ok {
        if inferredType == "void" {
            // error...
        }
        // inferredType is "" here, but no error is raised!
        tc.defineVariableWithMutability(varName, inferredType, decl.Mutable)
    }
}
```

The variable is registered with an empty type string, causing misleading errors later.

## Suggested Fix

In `checkVariableDeclaration`, after inferring the type, check if it's empty and the value is a multi-return function call:

```go
if declaredType == "" {
    inferredType, ok := tc.inferExpressionType(decl.Value)
    if ok {
        if inferredType == "void" {
            // existing error...
        }
        if inferredType == "" {
            // Check if it's a multi-return function
            if tc.isMultiReturnCall(decl.Value) {
                tc.addError(
                    errors.E3001,
                    "cannot assign multi-return function result to single variable; use multiple variables or discard with _",
                    decl.Name.Token.Line,
                    decl.Name.Token.Column,
                )
                return
            }
        }
        tc.defineVariableWithMutability(varName, inferredType, decl.Mutable)
    }
}
```

## Related Code

- `pkg/typechecker/typechecker.go` lines 1808-1826 (single-variable declaration handling)
- `pkg/typechecker/typechecker.go` lines 5130-5134 (multi-return type inference)

--- a/ERRORS.md
+++ b/ERRORS.md
@@ -376,4 +376,4 @@ Module-related warnings
 
 ## Summary
 
-**Total:** 257 error/warning codes
+**Total:** 258 error/warning codes

--- a/ERRORS.md
+++ b/ERRORS.md
@@ -92,6 +92,7 @@ Syntax and parsing errors
 | E2054 | when-strict-non-enum-case | #strict when requires explicit enum member values in cases |
 | E2055 | strict-invalid-target | #strict can only be applied to when statements |
 | E2056 | executable-at-file-scope | executable statement not allowed at file scope |
+| E2057 | invalid-interpolation-syntax | invalid string interpolation syntax |
 
 ## Type Errors (E3xxx)
 

--- a/ERRORS.md
+++ b/ERRORS.md
@@ -138,6 +138,7 @@ Type system errors
 | E3037 | invalid-private-usage | private modifier cannot be used here |
 | E3038 | void-type-not-allowed | 'void' is not a valid type |
 | E3039 | ensure-expects-call | ensure expects a function call |
+| E3040 | multi-return-to-single-var | cannot assign multiple return values to single variable |
 
 ## Reference Errors (E4xxx)
 

--- a/ERRORS.md
+++ b/ERRORS.md
@@ -375,4 +375,4 @@ Module-related warnings
 
 ## Summary
 
-**Total:** 256 error/warning codes
+**Total:** 257 error/warning codes

--- a/cmd/ez/commands.go
+++ b/cmd/ez/commands.go
@@ -1,8 +1,11 @@
 package main
 
 import (
+	"os"
+	"slices"
 	"strings"
 
+	"github.com/marshallburns/ez/pkg/stdlib"
 	"github.com/spf13/cobra"
 )
 
@@ -83,14 +86,26 @@ var rootCmd = &cobra.Command{
 	Use:     "ez [file.ez]",
 	Short:   "EZ Language Interpreter",
 	Long:    "A simple, interpreted, statically-typed programming language designed for clarity and ease of use.",
-	Args:    cobra.MaximumNArgs(1),
+	Args:    cobra.ArbitraryArgs,
 	Version: getVersionString(),
 	RunE: func(cmd *cobra.Command, args []string) error {
-		if len(args) == 1 && strings.HasSuffix(args[0], ".ez") {
-			runFile(args[0])
-			return nil
+		if len(args) == 0 {
+			return cmd.Help()
 		}
-		return cmd.Help()
+		if !strings.HasSuffix(args[0], ".ez") {
+			return cmd.Help()
+		}
+
+		programArgs := []string{os.Args[0], args[0]}
+		if cmd.ArgsLenAtDash() > 0 {
+			programArgs = slices.Concat(programArgs, args[cmd.ArgsLenAtDash():])
+		} else if len(args) > 1 {
+			programArgs = slices.Concat(programArgs, args[1:])
+		}
+		stdlib.CommandLineArgs = programArgs
+
+		runFile(args[0])
+		return nil
 	},
 }
 

--- a/integration-tests/fail/errors/E2057_invalid_interpolation_syntax.ez
+++ b/integration-tests/fail/errors/E2057_invalid_interpolation_syntax.ez
@@ -1,0 +1,9 @@
+/*
+ * Error Test: E2057 - invalid-interpolation-syntax
+ * Expected: "invalid interpolation syntax" or "$identifier"
+ */
+
+do main() {
+    temp file string = "test.ez"
+    temp s string = "$File: something"
+}

--- a/integration-tests/fail/errors/E3031_bare_function_name.ez
+++ b/integration-tests/fail/errors/E3031_bare_function_name.ez
@@ -1,0 +1,12 @@
+/*
+ * Error Test: E3031 - function-as-value
+ * Expected: "function" and "cannot be used as a value"
+ */
+
+do some_function() {
+    // Some function
+}
+
+do main() {
+    some_function
+}

--- a/pkg/errors/codes.go
+++ b/pkg/errors/codes.go
@@ -138,6 +138,7 @@ var (
 	E3037 = ErrorCode{"E3037", "invalid-private-usage", "private modifier cannot be used here"}
 	E3038 = ErrorCode{"E3038", "void-type-not-allowed", "'void' is not a valid type"}
 	E3039 = ErrorCode{"E3039", "ensure-expects-call", "ensure expects a function call"}
+	E3040 = ErrorCode{"E3040", "multi-return-to-single-var", "cannot assign multiple return values to single variable"}
 )
 
 // =============================================================================

--- a/pkg/errors/codes.go
+++ b/pkg/errors/codes.go
@@ -93,6 +93,7 @@ var (
 	E2054 = ErrorCode{"E2054", "when-strict-non-enum-case", "#strict when requires explicit enum member values in cases"}
 	E2055 = ErrorCode{"E2055", "strict-invalid-target", "#strict can only be applied to when statements"}
 	E2056 = ErrorCode{"E2056", "executable-at-file-scope", "executable statement not allowed at file scope"}
+	E2057 = ErrorCode{"E2057", "invalid-interpolation-syntax", "invalid string interpolation syntax"}
 )
 
 // =============================================================================
@@ -349,11 +350,11 @@ var (
 // These are errors unique to HTTP operations
 // =============================================================================
 var (
-	E14001 = ErrorCode{"E14001", "http-invalid-url",					"invalid URL"}
-	E14002 = ErrorCode{"E14002", "http-failed-request",       "request failed (network error)"}
-	E14003 = ErrorCode{"E14003", "http-timeout",							"timeout exceeded"}
-	E14004 = ErrorCode{"E14004", "http-invalid-method",				"invalid HTTP method"}
-	E14005 = ErrorCode{"E14005", "http-failed-url-decode",		"URL decode failed"}
+	E14001 = ErrorCode{"E14001", "http-invalid-url", "invalid URL"}
+	E14002 = ErrorCode{"E14002", "http-failed-request", "request failed (network error)"}
+	E14003 = ErrorCode{"E14003", "http-timeout", "timeout exceeded"}
+	E14004 = ErrorCode{"E14004", "http-invalid-method", "invalid HTTP method"}
+	E14005 = ErrorCode{"E14005", "http-failed-url-decode", "URL decode failed"}
 	E14006 = ErrorCode{"E14006", "http-failed-json-encoding", "JSON encoding failed"}
 )
 

--- a/pkg/interpreter/evaluator.go
+++ b/pkg/interpreter/evaluator.go
@@ -147,23 +147,23 @@ var globalEvalContext *EvalContext
 
 // validModules lists all available standard library modules
 var validModules = map[string]bool{
-	"std":     true, // Standard I/O functions (println, print, read_int)
-	"math":    true, // Math functions
-	"strings": true, // String utilities
-	"arrays":  true, // Array utilities
-	"maps":    true, // Map utilities
-	"time":    true, // Time functions
-	"io":      true, // File system and I/O operations
-	"os":      true, // Operating system and environment
-	"bytes":   true, // Binary data operations
-	"random":  true, // Random number generation
-	"json":    true, // JSON encoding/decoding
-	"binary":  true, // Binary encoding/decoding for integers and floats
+	"std":      true, // Standard I/O functions (println, print, read_int)
+	"math":     true, // Math functions
+	"strings":  true, // String utilities
+	"arrays":   true, // Array utilities
+	"maps":     true, // Map utilities
+	"time":     true, // Time functions
+	"io":       true, // File system and I/O operations
+	"os":       true, // Operating system and environment
+	"bytes":    true, // Binary data operations
+	"random":   true, // Random number generation
+	"json":     true, // JSON encoding/decoding
+	"binary":   true, // Binary encoding/decoding for integers and floats
 	"db":       true, // Simple key-value database
 	"uuid":     true, // UUID generation and validation
 	"encoding": true, // Base64, hex, and URL encoding/decoding
 	"crypto":   true, // Cryptographic hashing and secure random
-	"http":			true, // HTTP client for web requests
+	"http":     true, // HTTP client for web requests
 }
 
 // isValidModule checks if a module name is valid (either standard library or user-created)
@@ -866,6 +866,14 @@ func evalVariableDeclaration(node *ast.VariableDeclaration, env *Environment) Ob
 			return NIL
 		}
 
+		// Defensive check: multi-value return assigned to single variable
+		// The typechecker should catch this (E3040), but guard against regressions
+		if retVal, ok := val.(*ReturnValue); ok && len(retVal.Values) > 1 {
+			return newErrorWithLocation("E5012", node.Token.Line, node.Token.Column,
+				"cannot assign %d values to single variable; use tuple unpacking: a, b = func()",
+				len(retVal.Values))
+		}
+
 		// Validate type compatibility if a type is declared (single variable case)
 		if node.TypeName != "" {
 			// Check if declared type is an array type
@@ -1207,14 +1215,14 @@ func evalAssignment(node *ast.AssignmentStatement, env *Environment) Object {
 							if v.Value.Sign() < 0 || v.Value.Cmp(big.NewInt(255)) > 0 {
 								return newErrorWithLocation("E3025", node.Token.Line, node.Token.Column,
 									"byte value must be between 0 and 255: %s", v.Value.String())
-								}
-								val = &Byte{Value: uint8(v.Value.Int64())}
-							case *Byte:
-								// ok
-							default:
-								return newErrorWithLocation("E3025", node.Token.Line, node.Token.Column,
-									"cannot assign %s to byte variable", objectTypeToEZ(val))
 							}
+							val = &Byte{Value: uint8(v.Value.Int64())}
+						case *Byte:
+							// ok
+						default:
+							return newErrorWithLocation("E3025", node.Token.Line, node.Token.Column,
+								"cannot assign %s to byte variable", objectTypeToEZ(val))
+						}
 					}
 				}
 				ref.SetValue(val)
@@ -2933,7 +2941,7 @@ func applyFunction(fn Object, args []Object, line, col int) Object {
 		}
 
 		evaluated := Eval(fn.Body, extendedEnv)
-		
+
 		// Execute ensure statements before returning (LIFO order)
 		ensures := extendedEnv.ExecuteEnsures()
 		for _, ensureCall := range ensures {
@@ -2941,7 +2949,7 @@ func applyFunction(fn Object, args []Object, line, col int) Object {
 			// Note: We ignore errors from ensure statements to ensure cleanup always runs
 		}
 		extendedEnv.ClearEnsures()
-		
+
 		// Restore current file
 		if globalEvalContext != nil && fn.File != "" {
 			globalEvalContext.CurrentFile = oldFile

--- a/pkg/parser/parser.go
+++ b/pkg/parser/parser.go
@@ -2640,6 +2640,11 @@ func (p *Parser) parseStringValue() Expression {
 	// Process escape sequences like \n, \t, etc.
 	literal = processEscapeSequences(literal)
 
+	// Check for invalid interpolation patterns before processing
+	if invalidPattern := detectInvalidInterpolation(literal); invalidPattern != "" {
+		p.addEZError(errors.E2057, fmt.Sprintf("invalid interpolation syntax '%s'; use ${...} for string interpolation", invalidPattern), token)
+	}
+
 	// Check if string contains interpolation patterns ${...}
 	if !containsInterpolation(literal) {
 		return &StringValue{Token: token, Value: literal}
@@ -2738,6 +2743,41 @@ func containsInterpolation(s string) bool {
 		}
 	}
 	return false
+}
+
+// detectInvalidInterpolation checks for likely invalid interpolation syntax.
+// Returns the first invalid pattern found, or empty string if none.
+// Detects patterns like:
+//   - $identifier ($ followed by identifier chars but not {)
+//
+// Note: We intentionally don't flag {identifier} patterns because they have
+// too many false positives with JSON and other legitimate uses of braces.
+func detectInvalidInterpolation(s string) string {
+	for i := 0; i < len(s); i++ {
+		// Check for $identifier pattern ($ not followed by {)
+		if s[i] == '$' {
+			if i+1 < len(s) && s[i+1] != '{' && isIdentChar(s[i+1]) {
+				// Found $ followed by identifier character but not {
+				// Extract the identifier
+				end := i + 1
+				for end < len(s) && isIdentChar(s[end]) {
+					end++
+				}
+				return s[i:end]
+			}
+		}
+	}
+	return ""
+}
+
+// isIdentStartChar checks if a byte can start an identifier (letter or underscore)
+func isIdentStartChar(c byte) bool {
+	return (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '_'
+}
+
+// isIdentChar checks if a byte can be part of an identifier
+func isIdentChar(c byte) bool {
+	return isIdentStartChar(c) || (c >= '0' && c <= '9')
 }
 
 // processEscapeSequences converts escape sequences like \n, \t to actual characters

--- a/pkg/stdlib/builtins.go
+++ b/pkg/stdlib/builtins.go
@@ -76,6 +76,20 @@ func getEZTypeName(obj object.Object) string {
 		return "nil"
 	case *object.Function:
 		return "function"
+	case *object.ReturnValue:
+		// Handle multi-return values - show tuple type
+		if len(v.Values) == 0 {
+			return "void"
+		}
+		if len(v.Values) == 1 {
+			return getEZTypeName(v.Values[0])
+		}
+		// Multiple values - show as tuple
+		types := make([]string, len(v.Values))
+		for i, val := range v.Values {
+			types[i] = getEZTypeName(val)
+		}
+		return "(" + strings.Join(types, ", ") + ")"
 	default:
 		return string(obj.Type())
 	}

--- a/pkg/stdlib/os.go
+++ b/pkg/stdlib/os.go
@@ -136,13 +136,8 @@ var OSBuiltins = map[string]*object.Builtin{
 	"os.args": {
 		Fn: func(args ...object.Object) object.Object {
 			// Use CommandLineArgs if set, otherwise fall back to os.Args
-			sourceArgs := CommandLineArgs
-			if len(sourceArgs) == 0 {
-				sourceArgs = os.Args
-			}
-
-			elements := make([]object.Object, len(sourceArgs))
-			for i, arg := range sourceArgs {
+			elements := make([]object.Object, len(CommandLineArgs))
+			for i, arg := range CommandLineArgs {
 				elements[i] = &object.String{Value: arg}
 			}
 			return &object.Array{Elements: elements, Mutable: false, ElementType: "string"}

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -1818,6 +1818,18 @@ func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 					)
 					return
 				}
+				if inferredType == "" {
+					// Check if it's a multi-return function being assigned to a single variable
+					if tc.isMultiReturnCall(decl.Value) {
+						tc.addError(
+							errors.E3040,
+							"cannot assign multi-return function result to single variable; use multiple variables or discard with _",
+							decl.Name.Token.Line,
+							decl.Name.Token.Column,
+						)
+						return
+					}
+				}
 				tc.defineVariableWithMutability(varName, inferredType, decl.Mutable)
 			} else {
 				// Still register with empty type so variable is in scope
@@ -1970,6 +1982,60 @@ func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 	if declaredType != "" {
 		tc.defineVariableWithMutability(varName, declaredType, decl.Mutable)
 	}
+}
+
+// isMultiReturnCall checks if an expression is a function call that returns multiple values
+func (tc *TypeChecker) isMultiReturnCall(expr ast.Expression) bool {
+	callExpr, ok := expr.(*ast.CallExpression)
+	if !ok {
+		return false
+	}
+
+	// Get the function name and module name (if applicable)
+	var funcName string
+	var moduleName string
+
+	switch fn := callExpr.Function.(type) {
+	case *ast.Label:
+		funcName = fn.Value
+	case *ast.MemberExpression:
+		// Module.function call - get both module and function names
+		funcName = fn.Member.Value
+		if obj, ok := fn.Object.(*ast.Label); ok {
+			moduleName = obj.Value
+		}
+	default:
+		return false
+	}
+
+	// Look up the function signature
+	if funcSig, exists := tc.functions[funcName]; exists {
+		return len(funcSig.ReturnTypes) > 1
+	}
+
+	// Check if it's a module function with multiple return values
+	if moduleName != "" {
+		resolvedModuleName := tc.resolveStdlibModule(moduleName)
+		if returnTypes := tc.getModuleMultiReturnTypes(resolvedModuleName, funcName, callExpr.Arguments); returnTypes != nil {
+			return len(returnTypes) > 1
+		}
+	}
+
+	// Check if it's a function from a user-defined module via 'using'
+	for usingModuleName := range tc.fileUsingModules {
+		if moduleFuncs, hasModule := tc.moduleFunctions[usingModuleName]; hasModule {
+			if moduleSig, found := moduleFuncs[funcName]; found {
+				return len(moduleSig.ReturnTypes) > 1
+			}
+		}
+	}
+
+	// Check if it's a builtin function with multiple return values
+	if builtinReturnTypes := tc.getBuiltinMultiReturnTypes(funcName); builtinReturnTypes != nil {
+		return len(builtinReturnTypes) > 1
+	}
+
+	return false
 }
 
 // checkMultiReturnDeclaration validates multi-return variable declarations

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -1807,6 +1807,18 @@ func (tc *TypeChecker) checkVariableDeclaration(decl *ast.VariableDeclaration) {
 
 		// If no declared type, infer from value and register it
 		if declaredType == "" {
+			// Check if it's a multi-return function being assigned to a single variable
+			// Must check BEFORE type inference since inferCallType returns first type for multi-return
+			if tc.isMultiReturnCall(decl.Value) {
+				tc.addError(
+					errors.E3040,
+					"cannot assign multi-return function result to single variable; use multiple variables or discard with _",
+					decl.Name.Token.Line,
+					decl.Name.Token.Column,
+				)
+				return
+			}
+
 			inferredType, ok := tc.inferExpressionType(decl.Value)
 			if ok {
 				if inferredType == "void" {

--- a/pkg/typechecker/typechecker.go
+++ b/pkg/typechecker/typechecker.go
@@ -2513,6 +2513,10 @@ func (tc *TypeChecker) checkExpressionStatement(exprStmt *ast.ExpressionStatemen
 		return
 	}
 
+	// Check for bare identifiers that have no effect as statements
+	// A bare function name or type name as a statement does nothing
+	tc.checkValueExpression(exprStmt.Expression)
+
 	// Validate the entire expression tree
 	tc.checkExpression(exprStmt.Expression)
 }

